### PR TITLE
Add marker interface and store source of last change to MouseState

### PR DESCRIPTION
**Pull request by peppy**
Alternative to https://github.com/ppy/osu-framework/pull/3675/files using interface rather than polluting non-touch classes with touch-specific booleans. Took same test from other PR to make it easy to compare implementation / usage.

### CODE REVIEW

File: osu.Framework/Input/States/MouseState.cs [hunk 1]
```diff
@@ -17,6 +18,11 @@ public class MouseState
 
         public bool IsPositionValid { get; set; } = true;
 
+        /// <summary>
+        /// The last input source to make a change to state.
```
Comment (frenzibyte) [inline osu.Framework/Input/States/MouseState.cs:22]:
```suggestion
        /// The last input source making changes to this state.
```

File: osu.Framework/Input/StateChanges/ISourcedFromTouch.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Input.StateChanges.Events;
+
+namespace osu.Framework.Input.StateChanges
+{
+    public interface ISourcedFromTouch
+    {
+        TouchStateChangeEvent TouchEvent { get; set; }
+    }
```
Comment (bdach) [inline osu.Framework/Input/StateChanges/ISourcedFromTouch.cs:11]: I'd say this should have xmldoc.

**Comment by UselessToucan** [comment]
`ISourcedFromTouch`, `MouseButtonInputFromTouch`, `MouseButtonInputFromTouch` are probably unnecessary

see https://github.com/UselessToucan/osu-framework/commit/5c07165c246acb89abdc52372407deff3f5a69f1

**Comment by bdach** [comment]
Kind of out of the loop on the touch stuff so I'd rather abstain from reviewing this if you don't mind (I guess I can also try and get caught up but unsure if I will be able to pick up all the nuances).

File: osu.Framework/Input/StateChanges/ISourcedFromTouch.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    /// <summary>
+    /// Denotes an input which was sourced from a touch event.
+    /// Generally used to mark when an alternate input was triggered from a touch source (ie. touch being emulated as a mouse).
+    /// </summary>
+    public interface ISourcedFromTouch : IInput
+    {
+        /// <summary>
+        /// The source touch event.
+        /// </summary>
+        TouchStateChangeEvent TouchEvent { get; set; }
```
Comment (smoogipoo) [inline osu.Framework/Input/StateChanges/ISourcedFromTouch.cs:17]: Maybe `get;` only and add to ctor? This shouldn't be settable from the outside. This is also true for all other inputs, but this is more likely the one that's going to be used if any.

**Comment by peppy** [comment]
@smoogipoo Yes I looked at the comment and disagree with it. I think you may agree if you check the required changes (it looks to add complexity and reduce readability).

**Comment by smoogipoo** [comment]
I'd already looked at it by the time I posted that comment. What part increases complexity and reduces readability? I see it as reducing complexity because there's not some arbitrary marker interface somewhere that is pretty new to the structure here.

At least state change events are [documented](https://github.com/ppy/osu-framework/wiki/Diagram:-From-input-handlers-to-events), minus touch and midi.

**Comment by peppy** [comment]
Reading that diagram makes the proposed change even less understandable. It's associating a `ChangeEvent` with an `IInput`, but `ChangeEvents` are the result of an `IInput`? "An event that caused this input to be created" I don't think anyone would be able to understand this.

**Comment by smoogipoo** [comment]
I see, thanks.

**Pull request merged**